docs: Custom cosmjs interfaces tutorial update (#38)
cosmos/developer-portal: tutorials/7-cosmjs/5-create-custom.md

## Compiling the Protobuf objects and clients

> You can choose which library you use to compile your Protobuf objects into TypeScript or JavaScript. Reproducing what [cosmjs-types](https://github.com/confio/cosmjs-types/blob/main/scripts/codegen.js) or [Stargate](https://github.com/cosmos/cosmjs/blob/main/packages/stargate/CUSTOM_PROTOBUF_CODECS.md) do is a good choice.

Reviewer: Why did you swap the order?

Author: I just thought since we're telling users how to use telescope way of compilation in tutorial and cosmjs's using telescope to generate, so how about we switch the order and put cosmjs sample code ahead as an emphasis. If you think we still keep the order I'll change it back.

$ npm install ts-proto@1.121.6 --save-dev
```

Adjust to your preferred version, operating system, and CPU platform. For instance, on an Apple M1 you would use `protoc-21.7-osx-aarch_64.zip`.

</CodeGroupItem>

<CodeGroupItem title="Dockerfile">

```Dockerfile
FROM --platform=linux node:lts-slim as base
ARG BUILDARCH

ENV PROTOC_VERSION=21.7
ENV TS_PROTO_VERSION=1.121.6

FROM base AS platform-amd64
ENV PROTOC_PLATFORM=x86_64

FROM base AS platform-arm64
ENV PROTOC_PLATFORM=aarch_64

FROM platform-${BUILDARCH}

RUN apt-get update
RUN apt-get install curl unzip --yes

# Install ProtoC
RUN mkdir -p /usr/lib/protoc
WORKDIR /usr/lib/protoc
RUN curl -L https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-${PROTOC_PLATFORM}.zip -o protoc.zip
RUN unzip -o protoc.zip
RUN rm protoc.zip
RUN ln -s /usr/lib/protoc/bin/protoc /usr/local/bin/protoc

# Install ts-proto
RUN npm install --global ts-proto@${TS_PROTO_VERSION} --save-exact

WORKDIR /

ENTRYPOINT [ "protoc" ]
$ npm install --save-dev --save-exact @cosmology/telescope@1.0.4
```

Then build the image:

```sh
$ docker build . -t ts-protoc
```

</CodeGroupItem>

> </CodeGroup>

Reviewer: Can you keep the CodeGroup like you did for 1-cosmjs-objects

---

You can confirm the version you received. The executable is located in `./node_modules/protoc/protoc/bin/protoc`:

<CodeGroup>

<CodeGroupItem title="Local">
You can confirm the version you received by running:

```sh
$ protoc --version
$ npx telescope --version
```

</CodeGroupItem>

<CodeGroupItem title="Docker">

```sh
$ docker run --rm -it \
    ts-protoc --version
```

</CodeGroupItem>

> </CodeGroup>

Reviewer: Here too keep the CodeGroup.

This returns something like:

```txt
libprotoc 3.21.7
1.0.4
```

The compiler tools are ready. Time to use them.

Create the target folder if it does not exist yet:

```sh

> $ mkdir -p client/src/types/generated

Reviewer: Is the folder created automatically?

Author: Yes, outPath will be created by Telescope itself

Reviewer: Maybe we can mention that as part of the notes.

Perhaps as a troubleshoot...In particular, auto-creation of folders usually comes with silent errors, i.e. the path was wrongly written, the command completed but the user has no idea where everything went...

Author: This's a very good point, I've tried to make it up.

2. Your Protobuf definition files are in `./proto/`.
3. You want to transpile them into TypeScript in `./src/codegen/`.

Install `telescope`, in your package or in docker:

<CodeGroup>

<CodeGroupItem title="Local">

```sh
$ npm install --save-dev --save-exact @cosmology/telescope@1.0.4
```

</CodeGroupItem>

<CodeGroupItem title="Docker">

```sh
$ docker run --rm -it \
    -v $(pwd):/myLib \
    -w /myLib \
    checkers_i \
    npm install --save-dev --save-exact @cosmology/telescope@1.0.4
```

</CodeGroupItem>

> <CodeGroupItem title="DockerGlobal">

Reviewer: ```suggestion
<CodeGroupItem title="Docker Image">
```

```sh
# Install Node
RUN curl -fsSL https://deb.nodesource.com/setup_${NODE_VERSION} | bash -

> RUN apt-get install -y nodejs

Reviewer: What about:

```
FROM node:x.y
```

Like this you also make explicit which Node version will work.

# Add this line after installing Node to install Telescope in your dockerfile

> RUN npm install --global @cosmology/telescope@1.0.4

Reviewer: ```suggestion
RUN npm install --global @cosmology/telescope@1.0.4

ENTRYPOINT ["npx", "telescope"]
```

Perhaps, no?

```

</CodeGroupItem>

</CodeGroup>

---

You can confirm the version you received by running:

<CodeGroup>

<CodeGroupItem title="Local">

```sh
$ npx telescope --version
```

</CodeGroupItem>

> <CodeGroupItem title="Local">

Reviewer: ```suggestion
<CodeGroupItem title="Docker">
```

```sh
$ docker run --rm -it \
    -v $(pwd):/myLib \
    -w /myLib \
    checkers_i \
    npx telescope --version
```

</CodeGroupItem>

> 

Reviewer: You could add a code group item that uses the new image.

$ curl https://raw.githubusercontent.com/cosmos/cosmos-sdk/v0.45.4/third_party/proto/gogoproto/gogo.proto -o ./proto/gogoproto/gogo.proto
```

Or you can add latest whole `third_party` folders by running `npx telescope install` under `myLib` folder:

<CodeGroup>
<CodeGroupItem title="Interactive">

```sh
$ npx telescope install
```

then select packages you want to include in `myLib/proto` folder:

```txt
Telescope 1.0.3
? [pkg] which packages do you want to support? (Press <space> to select, <a> to toggle all, <i> to invert selection)
❯◯ amino
 ◯ akash
 ◯ bcna
 ...
 ◯ cosmos
 ...
```

Say `cosmos`'s selected and entered, there'll be cosmos folder and its dependencies(like gogoproto and google, etc) in `myLib/proto` folder.

> </CodeGroupItem>

Reviewer: ```suggestion

</CodeGroupItem>
```

<CodeGroupItem title="NonInteractive">

You can add wanted proto packages after `telescope install`, then you don't have to select them one by one.

```sh

> $ npx telescope install @protobufs/cosmos@0.0.10  @protobufs/confio

Reviewer: And add a Docker non-interactive?

```

Then there'll be the packages and its dependencies(like gogoproto and google, etc) in myLib/proto folder.

</CodeGroupItem>
</CodeGroup>

### Compilation

You can now compile the Protobuf files using telescope.

```sh

> $ npx telescope transpile

Reviewer: Again, have a Docker one.

```

You should now see some `.ts` files generated in `./src/codegen`. These are the real source files used in your application.

By default, the command takes `./proto` as input folder and `./src/codegen` as output folder. You can config in and out folders and the way telescope generating code by using options like the examples:

```sh
# Telescope takes chain1 folder as input,
# and generate files in 'gen/src' folder.
$ npx telescope transpile --protoDirs ./proto --outPath gen/src
```

Each time `telescope transpile` has been ran, a config file `.telescope.json` will be created in the folder where it's running:

```json
//.telescope.json
{
  "protoDirs": [
    "../../proto/chain2"
  ],
  "outPath": "gen/src",
  "options": {
    // telescope options
    "addTypeUrlToDecoders": true,
    "addTypeUrlToObjects": true,

    "typingsFormat": {
        "num64": "bigint",
        "customTypes": {
            "useCosmosSDKDec": true
        }
    },
    ...
  }
}
```

> You can modify the config file according to [telescope options](https://github.com/cosmology-tech/telescope#options), and then use the file by the option `--config`:

Reviewer: This link could be added in `1-cosmjs-objects` too.

Author: Good catch! Done!

```sh
# Telescope takes chain1 folder(from args) and chain2 folder(from config) as input,

> # and generate files in 'gen/src'(defined in the config file, will override outPath in args) folder using a config file.

Reviewer: This comment can go out of the code block.

Then there'll be the packages and its dependencies(like gogoproto and google, etc) in myLib/proto folder.

</CodeGroupItem>

> <CodeGroupItem title="DockerNonInteractive">

Reviewer: ```suggestion
<CodeGroupItem title="Docker non-interactive">
```

}
```

> If so, you find its service declaration in the compiled `tx.ts` file(with telescope option `rpcClients.inline: true`) or in `tx.rpc.msg.ts` file(with telescope option `rpcClients.inline: false`):

Reviewer: ```suggestion
If so, you find its service declaration in the compiled `tx.ts` file (with telescope option `rpcClients.inline: true`) or in `tx.rpc.msg.ts` file(with telescope option `rpcClients.inline: false`):
```

and everywhere else you open parentheses.

> ```typescript [https://github.com/confio/cosmjs-types/blob/v0.9.0-rc.1/src/cosmos/bank/v1beta1/tx.ts#L484]

Reviewer: Could it be a non "rc-1" version?

You can modify the config file according to [telescope options](https://github.com/cosmology-tech/telescope#options), and then use the file by the option `--config`:

<CodeGroup>
<CodeGroupItem title="Local">

```sh
$ npx telescope transpile --protoDirs ../../proto-common --config .telescope.json
```

</CodeGroupItem>
<CodeGroupItem title="Docker">

```sh
$ docker run --rm -it \
    -v $(pwd):/myLib \
    -w /myLib \
    checkers_i \
    npx telescope transpile --protoDirs ../../proto-common --config .telescope.json
```

</CodeGroupItem>
</CodeGroup>

> When running above command, telescope takes proto-common folder(from args) and proto folder(from config) as input, and generate files in 'gen/src'(defined in the config file, will override outPath in args) folder using a config file.

Reviewer: ```suggestion
When running the above command, telescope takes the `proto-common` folder (from arguments) and `proto` folder (from the configuration file) as inputs, and generates files in `gen/src` (defined as `outPath` in the configuration file, which overrides `outPath` if present in arguments) folder using a configuration file.
```

Still a bit confusing. Perhaps bullet points?

Author: Thanks! I tried to make it more clear!

After it's been ran successfully, there'll be a message:

```txt
✨ files transpiled in '/gen/src'
✨ transpilation successful!
```

<HighlightBox type="note">

> Note: `outPath` will be automatically created after transpilation's successful

Reviewer: ```suggestion
Note that `outPath` is created automatically during the transpilation. This means that you will get no "missing dir" error, for instance if your `outPath` is incorrect.
```
